test/oqs_test_libctx.c

```diff
@@ -0,0 +1,563 @@
+    size_t pubkey1_len, pubkey2_len, privkey1_len, privkey2_len;
+
+    if (get_param_octet_string(key1, OSSL_PKEY_PARAM_PUB_KEY, &pubkey1,
+                               &pubkey1_len)) {
+        goto out;
+    }
+    if (get_param_octet_string(key2, OSSL_PKEY_PARAM_PUB_KEY, &pubkey2,
+                               &pubkey2_len)) {
+        goto out;
+    }
+
+    if (get_param_octet_string(key1, OSSL_PKEY_PARAM_PRIV_KEY, &privkey1,
+                               &privkey1_len)) {
+        goto out;
+    }
+    if (get_param_octet_string(key2, OSSL_PKEY_PARAM_PRIV_KEY, &privkey2,
+                               &privkey2_len)) {
+        goto out;
+    }
+
+    if ((idx = oqs_get_idx(sigalg_name)) >= OSSL_NELEM(info_classical)) {
+        goto out;
+    }
+
+    if (!((pubkey1_len == pubkey2_len) && (privkey1_len == privkey2_len))) {
+        goto out;
+    }
+
+    ret = memcmp(pubkey1, pubkey2, info_classical[idx].pubkey_len) == 0 &&
+          memcmp(privkey1, privkey2, info_classical[idx].privkey_len) == 0;
+
+    ret_rev = memcmp(pubkey1 + pubkey1_len - info_classical[idx].pubkey_len,
+                     pubkey2 + pubkey2_len - info_classical[idx].pubkey_len,
+                     info_classical[idx].pubkey_len) == 0 &&
+              memcmp(privkey1 + privkey1_len - info_classical[idx].privkey_len,
+                     privkey2 + privkey2_len - info_classical[idx].privkey_len,
+                     info_classical[idx].privkey_len) == 0;
+
+out:
+    free(pubkey1);
+    free(pubkey2);
+    free(privkey1);
+    free(privkey2);
+    return ret || ret_rev;
+}
+
+/** \brief Compares the classical KEM elemets of two Encaps/Decaps executions
+ *
+ * \param kemalg_name algorithm name.
+ * \param sec1 A shared secret.
+ * \param sec2 A shared secret.
+ * \param seclen Shared secret length.
+ * \param ct1 An encapsulation.
+ * \param ct1 An encapsulation.
+ * \param ctlen Encapsulation lentgh.
+ *
+ * \return 1 if the compare operation is successful, else 0. */
+static int oqs_cmp_kem_elems(const char *kemalg_name, const unsigned char *sec1,
+                             const unsigned char *sec2, size_t seclen,
+                             const unsigned char *ct1, const unsigned char *ct2,
+                             size_t ctlen) {
+    int ret = 1, ret_rev = 1, idx = 0;
+
+    if ((idx = oqs_get_idx(kemalg_name)) >= OSSL_NELEM(info_classical)) {
+        return 0;
+    }
+
+    ret_rev &= memcmp(sec1 + seclen - info_classical[idx].sec_len,
+                      sec2 + seclen - info_classical[idx].sec_len,
+                      info_classical[idx].sec_len) == 0 &&
+               memcmp(ct1 + ctlen - info_classical[idx].pubkey_len,
+                      ct2 + ctlen - info_classical[idx].pubkey_len,
+                      info_classical[idx].pubkey_len) == 0;
+
+    ret &= memcmp(sec1, sec2, info_classical[idx].sec_len) == 0 &&
+           memcmp(ct1, ct2, info_classical[idx].pubkey_len) == 0;
+
+    return ret || ret_rev;
```

Comment:
nit: why separate variables?

Reply:
I was separating the check of the buffer and its reverse check, but I have grouped into one variable (and avoid the two checks when possible).